### PR TITLE
[UWP] Enhanced background image support

### PR DESCRIPTION
**Pull request by carlos-zamora**
Allows background image to be repeated along the x and/or y axis. Also added background image support to containers and columns. Main changes were made to SharedModel and UWP. Android and iOS were just patched.

PR for #1291 
Spec: #477 

# SharedModel
- Added `BackgroundImage.h/cpp` for new BackgroundImage object
- Updated EverythingBagel with proper BackgroundImage validation
- Added a few new Adaptive Card constructors
- Added BackgroundImage to column/container

# UWP
- Added `AdaptiveBackgroundImage` and necessary changes to take in BackgroundImage as an object
- Added BackgroundImage to column/container
- Removed ambiguity for `Windows::UI::Xaml::VerticalAlignment` and `AdaptiveNamespace::VerticalAlignment` (same for HorizontalAlignment)
- `TileControl` (BIG CHANGE!!)
  - basically figure out number and location of tiles when background image is set to repeat, repeatVertically, or repeatHorizontally
  - tricky part is that it has to wait for the image to be loaded so that the dimensions of the image can be accessed

# Testing
Added 5 tests:
- `BackgroundImage.Repeat.json`
- `BackgroundImage.RepeatHorizontally.Center.json`
- `BackgroundImage.RepeatVertically.Center.json`
- `Column.BackgroundImage.json`
- `Container.BackgroundImage.json`

Note: there are no tests for `AdaptiveCard.BackgroundImage` where `RepeatHorizontally` and `RepeatVertically` have non-center alignment. This was done as a part of reducing TestApp payload.

### CODE REVIEW

**Comment by carlos-zamora** [comment]
@RebeccaAnne For some reason, the UWPTestApp is failing on most of the cards. But when I use the visualizer on some of the cards, I don't get that problem. I also don't get this problem in master so it's clearly something I did. Could you take a look please?

**Comment by khouzam** [comment]
This is failing:     DEFINE_ADAPTIVECARD_ENUM_DEFAULT(BackgroundImageMode, BackgroundImageMode::Stretch, {
        { BackgroundImageMode::Stretch, "stretch" },
        { BackgroundImageMode::RepeatHorizontally, "repeatHorizontally" },
        { BackgroundImageMode::RepeatVertically, "repeatVertically" },
        { BackgroundImageMode::Repeat, "repeat" }});




---
In reply to: [455678272](https://github.com/Microsoft/AdaptiveCards/pull/2316#issuecomment-455678272) [](ancestors = 455678272)

**Comment by paulcam206** [comment]
Looks like `AdaptiveCardSchemaKey` has entries for `Repeat`, `RepeatHorizontally`, and `RepeatVertically` in `Enums.h`, but no corresponding entries in the definition in `Enums.cpp`

---
In reply to: [455695884](https://github.com/Microsoft/AdaptiveCards/pull/2316#issuecomment-455695884) [](ancestors = 455695884,455678272)

**Comment by khouzam** [comment]
MakeAndInitialize on AdaptiveBackgroundImage(image) checks for nullptr, but doesn't handle the case where image is null, so the values are never initialized.

Also, the serialization is slightly wrong. When BackgroundImage is the default, it shouldn't be serialized at all, not it serializes to null.

---
In reply to: [455700696](https://github.com/Microsoft/AdaptiveCards/pull/2316#issuecomment-455700696) [](ancestors = 455700696,455695884,455678272)

File: source/uwp/Renderer/lib/AdaptiveCard.cpp [hunk 1]
```diff
@@ -194,14 +195,15 @@ namespace AdaptiveNamespace
         return S_OK;
     }
 
-    HRESULT AdaptiveCard::get_BackgroundImage(_Outptr_ HSTRING* backgroundImage)
+    _Use_decl_annotations_ HRESULT AdaptiveCard::get_BackgroundImage(IAdaptiveBackgroundImage** backgroundImage)
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/AdaptiveCard.cpp]: >_Use_decl_annotations_ [](start = 4, length = 22)

Don't use _Use_decl_annotations_, add the annotations to both the declaration and the definition

File: source/uwp/Renderer/lib/AdaptiveColumn.h [hunk 2]
```diff
@@ -41,6 +42,9 @@ namespace AdaptiveNamespace
         IFACEMETHODIMP get_VerticalContentAlignment(_Out_ ABI::AdaptiveNamespace::VerticalContentAlignment* verticalAlignment);
         IFACEMETHODIMP put_VerticalContentAlignment(ABI::AdaptiveNamespace::VerticalContentAlignment verticalAlignment);
 
+        IFACEMETHODIMP get_BackgroundImage(_Out_ ABI::AdaptiveNamespace::IAdaptiveBackgroundImage** backgroundImage);
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/AdaptiveColumn.h]: >_Out_ [](start = 43, length = 5)

Outptr

File: source/uwp/Renderer/idl/AdaptiveCards.Rendering.Uwp.idl [hunk 3]
```diff
@@ -738,24 +769,60 @@ AdaptiveNamespaceStart
...
+#endif
     ]
     runtimeclass AdaptiveMedia
     {
         [default] interface IAdaptiveMedia;
         interface IAdaptiveCardElement;
     };
 
+    [
+        version(NTDDI_WIN10_RS1),
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/idl/AdaptiveCards.Rendering.Uwp.idl]: Make this exclusiveto AdaptiveBackgroundImage

File: source/uwp/Renderer/lib/XamlBuilder.h [hunk 4]
```diff
@@ -47,11 +47,11 @@ namespace AdaptiveNamespace
                                    _In_ ABI::AdaptiveNamespace::IAdaptiveRenderContext* renderContext,
                                    _In_ ABI::AdaptiveNamespace::IAdaptiveRenderArgs* renderArgs,
                                    _Outptr_ ABI::Windows::UI::Xaml::IUIElement** textBlockControl);
-        static void BuildContainer(_In_ ABI::AdaptiveNamespace::IAdaptiveCardElement* adaptiveCardElement,
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/XamlBuilder.h]: >static [](start = 8, length = 6)

I'd prefer we keep these static if we can. What did you need that required making this non-static?

File: source/uwp/Renderer/lib/AdaptiveContainer.cpp [hunk 5]
```diff
@@ -87,6 +89,17 @@ namespace AdaptiveNamespace
         return S_OK;
     }
 
+    _Use_decl_annotations_ HRESULT AdaptiveContainer::get_BackgroundImage(IAdaptiveBackgroundImage** backgroundImage)
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/AdaptiveContainer.cpp]: >_Use_decl_annotations_ [](start = 4, length = 22)

Remove _Use_decl_annotations_

File: source/uwp/Renderer/lib/AdaptiveContainer.h [hunk 6]
```diff
@@ -35,6 +36,9 @@ namespace AdaptiveNamespace
         IFACEMETHODIMP get_VerticalContentAlignment(_Out_ ABI::AdaptiveNamespace::VerticalContentAlignment* verticalAlignment);
         IFACEMETHODIMP put_VerticalContentAlignment(ABI::AdaptiveNamespace::VerticalContentAlignment verticalAlignment);
 
+        IFACEMETHODIMP get_BackgroundImage(_Out_ ABI::AdaptiveNamespace::IAdaptiveBackgroundImage** backgroundImage);
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/AdaptiveContainer.h]: >_Out_ [](start = 43, length = 5)

outptr

File: source/uwp/Renderer/lib/AdaptiveImage.h [hunk 7]
```diff
@@ -44,8 +44,8 @@ namespace AdaptiveNamespace
         IFACEMETHODIMP get_AltText(_Outptr_ HSTRING* text);
         IFACEMETHODIMP put_AltText(_In_ HSTRING text);
 
-        IFACEMETHODIMP get_HorizontalAlignment(_Out_ ABI::AdaptiveNamespace::HAlignment* hAlignment);
-        IFACEMETHODIMP put_HorizontalAlignment(ABI::AdaptiveNamespace::HAlignment hAlignment);
+        IFACEMETHODIMP get_HorizontalAlignment(_Out_ ABI::AdaptiveNamespace::HorizontalAlignment* HorizontalAlignment);
+        IFACEMETHODIMP put_HorizontalAlignment(_In_ ABI::AdaptiveNamespace::HorizontalAlignment HorizontalAlignment);
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/AdaptiveImage.h]: >_In_ [](start = 47, length = 4)

You don't need an annotation here

File: source/uwp/Renderer/lib/AdaptiveTextBlock.cpp [hunk 8]
```diff
@@ -121,13 +121,13 @@ namespace AdaptiveNamespace
         return S_OK;
     }
 
-    HRESULT AdaptiveTextBlock::get_HorizontalAlignment(_Out_ ABI::AdaptiveNamespace::HAlignment* alignment)
+    HRESULT AdaptiveTextBlock::get_HorizontalAlignment(ABI::AdaptiveNamespace::HorizontalAlignment* alignment)
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/AdaptiveTextBlock.cpp]: Add out annotation

File: source/uwp/Renderer/lib/TileControl.h [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+                                              Microsoft::WRL::ComposableBase<ABI::Windows::UI::Xaml::Controls::IContentControlFactory>>
+    {
+        AdaptiveRuntimeStringClass(TileControl);
+
+    public:
+        HRESULT RuntimeClassInitialize() noexcept;
+
+        virtual HRESULT STDMETHODCALLTYPE put_BackgroundImage(_In_ IAdaptiveBackgroundImage* value);
+        virtual HRESULT STDMETHODCALLTYPE put_RootElement(_In_ IFrameworkElement* value);
+        virtual HRESULT STDMETHODCALLTYPE get_ResolvedImage(_In_ IUIElement** value);
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/TileControl.h]: >_In_ [](start = 60, length = 4)

Outptr

File: source/uwp/Renderer/lib/TileControl.h [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        virtual HRESULT STDMETHODCALLTYPE put_ImageSize(_In_ ABI::Windows::Foundation::Size value);
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/TileControl.h]: >_In_ [](start = 56, length = 4)

remove annotation

File: source/uwp/Renderer/lib/TileControl.h [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+        virtual HRESULT STDMETHODCALLTYPE LoadImageBrush(_In_ IUIElement* image);
+
+        // IFrameworkElementOverrides overrides
+        virtual HRESULT STDMETHODCALLTYPE OnApplyTemplate();
+        virtual HRESULT STDMETHODCALLTYPE MeasureOverride(_In_ Size availableSize, _Out_ Size* pReturnValue);
+        virtual HRESULT STDMETHODCALLTYPE ArrangeOverride(_In_ Size arrangeBounds, _Out_ Size* pReturnValue);
+
+        // not implemented
+        virtual HRESULT STDMETHODCALLTYPE get_BackgroundImage(_In_ IAdaptiveBackgroundImage** value) { return E_NOTIMPL; }
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/TileControl.h]: >_In_ [](start = 62, length = 4)

Outptr

File: source/uwp/Renderer/lib/TileControl.h [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+        virtual HRESULT STDMETHODCALLTYPE get_RootElement(_In_ IFrameworkElement** value) { return E_NOTIMPL; }
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/TileControl.h]: >_In_ [](start = 58, length = 4)

Outptr

File: source/uwp/Renderer/lib/TileControl.h [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+        virtual HRESULT STDMETHODCALLTYPE put_ResolvedImage(_In_ IUIElement* value) { return E_NOTIMPL; }
+        virtual HRESULT STDMETHODCALLTYPE get_ImageSize(_In_ ABI::Windows::Foundation::Size* value) { return E_NOTIMPL; }
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/TileControl.h]: >_In_ [](start = 56, length = 4)

Out

File: source/uwp/Renderer/lib/TileControl.h [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+        // ITypePeek method
+        void* PeekAt(REFIID riid) override { return PeekHelper(riid, this); }
+
+    private:
+        void RefreshContainerTile();
+        HRESULT ExtractBackgroundImageData(ABI::AdaptiveNamespace::BackgroundImageMode* mode,
+                                           ABI::AdaptiveNamespace::HorizontalAlignment* hAlignment,
+                                           ABI::AdaptiveNamespace::VerticalAlignment* vAlignment);
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/TileControl.h]: Annotations

File: source/uwp/Renderer/lib/TileControl.h [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+using namespace ABI::Windows::UI::Xaml;
+using namespace ABI::Windows::UI::Xaml::Shapes;
+using namespace ABI::Windows::UI::Xaml::Controls;
+using namespace ABI::Windows::UI::Xaml::Media;
+
+namespace AdaptiveNamespace
+{
+    class DECLSPEC_UUID("0F485063-EF2A-400E-A946-73E00EDFAC83") TileControl
+        : public Microsoft::WRL::RuntimeClass<ABI::AdaptiveNamespace::ITileControl,
+                                              //ABI::Windows::UI::Xaml::Controls::IContentControl,
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/TileControl.h]: >//ABI::Windows::UI::Xaml::Controls::IContentControl, [](start = 46, length = 52)

Remove this

File: source/uwp/Renderer/lib/TileControl.cpp [hunk 16]
```diff
@@ -0,0 +1,319 @@
...
+        {
+            ComPtr<IImageSource> imageSource;
+            THROW_IF_FAILED(image->get_Source(&imageSource));
+
+            EventRegistrationToken eventToken;
+            ComPtr<IBitmapImage> bitmapImage;
+            THROW_IF_FAILED(imageSource.As(&bitmapImage));
+            THROW_IF_FAILED(bitmapImage->add_ImageOpened(Callback<IRoutedEventHandler>(
+                                                             [&](IInspectable* /*sender*/, IRoutedEventArgs * /*args*/) -> HRESULT {
+                                                                 ComPtr<IUIElement> uiElement;
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/TileControl.cpp]: >                                                      [](start = 12, length = 53)

@paulcam206, can you tell what's going on here that makes the auto-formatting so indented? Our other lambdas don't do this, but i'm not sure why this one is different.

File: source/uwp/Renderer/lib/TileControl.cpp [hunk 17]
```diff
@@ -0,0 +1,319 @@
...
+        }
+        return E_INVALIDARG;
+    }
+
+    HRESULT TileControl::OnApplyTemplate()
+    {
+        // Call OnApplyTemplate() for composable base
+        ComPtr<IFrameworkElementOverrides> base;
+        RETURN_IF_FAILED(GetComposableBase()->QueryInterface(__uuidof(IFrameworkElementOverrides),
+                                                             reinterpret_cast<void**>(base.GetAddressOf())));
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/TileControl.cpp]: You can use the .As function of ComPtr to do the QI:

        ComPtr<IInspectable> composableBase = GetComposableBase();
        ComPtr<IFrameworkElementOverrides> frameworkOverrides;
        RETURN_IF_FAILED(composableBase.As(&frameworkOverrides));

        return frameworkOverrides->OnApplyTemplate();

File: source/uwp/Renderer/lib/XamlBuilder.cpp [hunk 18]
```diff
@@ -455,35 +458,63 @@ namespace AdaptiveNamespace
...
                                             _In_ IAdaptiveRenderContext* renderContext,
                                             _In_ IAdaptiveRenderArgs* renderArgs)
     {
         // In order to reuse the image creation code paths, we simply create an adaptive card
         // image element and then build that into xaml and apply to the root.
         ComPtr<IAdaptiveImage> adaptiveImage;
+        HSTRING url;
         THROW_IF_FAILED(MakeAndInitialize<AdaptiveImage>(&adaptiveImage));
+        backgroundImage->get_Url(&url);
         adaptiveImage->put_Url(url);
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/XamlBuilder.cpp]: THROW_IF_FAILED() around both of these

File: source/uwp/Renderer/lib/XamlBuilder.cpp [hunk 19]
```diff
@@ -455,35 +458,63 @@ namespace AdaptiveNamespace
...
+        switch (mode)
+        {
+        case ABI::AdaptiveNamespace::BackgroundImageMode::Stretch:
+            // Ignored: horizontalAlignment, verticalAlignment
+            THROW_IF_FAILED(xamlImage->put_Stretch(Stretch::Stretch_UniformToFill));
+
+            THROW_IF_FAILED(xamlImage.As(&backgroundAsFrameworkElement));
+            THROW_IF_FAILED(backgroundAsFrameworkElement->put_VerticalAlignment(VerticalAlignment_Stretch));
+            break;
+        default:
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/XamlBuilder.cpp]: >default: [](start = 8, length = 8)

Make this explicit so we can see what values this is handling:

        case ABI::AdaptiveNamespace::BackgroundImageMode::RepeatHorizontally:
        case ABI::AdaptiveNamespace::BackgroundImageMode::RepeatVertically:
        case ABI::AdaptiveNamespace::BackgroundImageMode::Repeat:

File: source/uwp/Renderer/lib/XamlBuilder.h (hunk 4 above)
Comment (carlos-zamora) [reply source/uwp/Renderer/lib/XamlBuilder.h]: `XamlBuilder::BuildContainer()` and `XamlBuilder::BuildColumn()` now call `XamlBuilder::ApplyBackgroundToRoot()` (non-static).

I noticed a somewhat related issue, though. `XamlBuilder::BuildPanelChildren()` calls `AdaptiveContainerRenderer::Render()` which just calls `XamlBuilder::BuildContainer()`. Originally, I created a new instance of `XamlBuilder` to be able to do this. I'm realizing that instead I should be using the main/only instance of it.

Comment (RebeccaAnne) [reply source/uwp/Renderer/lib/XamlBuilder.h]: My ActionSet change (now in master) updated ApplyBackgroundToRoot to be static. Try merging that in and see if you can keep BuildContainer and BuildColumn static.

---
In reply to: [249991803](https://github.com/Microsoft/AdaptiveCards/pull/2316#discussion_r249991803) [](ancestors = 249991803)

File: samples/v1.2/Elements/BackgroundImage.Repeat.json [hunk 20]
```diff
@@ -0,0 +1,57 @@
+{
```
Comment (matthidinger) [inline samples/v1.2/Elements/BackgroundImage.Repeat.json]: Can these files follow similar naming convention for other element properties? E.g., `AdaptiveCard.BackgroundImage.Repeat.json`

**Comment by carlos-zamora** [comment]
Still have to run TestApp. Will do after it gets somewhat fixed on master.

File: samples/v1.2/Elements/AdaptiveCard.BackgroundImage.RepeatVertically.Center.json [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+		"horizontalAlignment": "center"
+	},
+  "body": [
+    {
+      "type": "TextBlock",
+      "text": "Text in body"
+    },
+    {
+      "type": "ColumnSet",
+      "columns": [
```
Comment (khouzam) [inline samples/v1.2/Elements/AdaptiveCard.BackgroundImage.RepeatVertically.Center.json]: Tabs please

File: source/uwp/Renderer/lib/AdaptiveBackgroundImage.cpp [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+            std::make_shared<AdaptiveSharedNamespace::BackgroundImage>();
+        return RuntimeClassInitialize(image);
+    }
+    CATCH_RETURN;
+
+    _Use_decl_annotations_ HRESULT AdaptiveBackgroundImage::RuntimeClassInitialize(
+        const std::shared_ptr<AdaptiveSharedNamespace::BackgroundImage>& sharedImage) try
+    {
+        std::shared_ptr<AdaptiveSharedNamespace::BackgroundImage> image = sharedImage;
+        if (sharedImage == nullptr)
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/AdaptiveBackgroundImage.cpp]: >if (sharedImage == nullptr) [](start = 8, length = 27)

Did you hit this case? Looks like our other types just return E_INVALIDARG, and while this seems fine, i'm wondering if it should be consistent

File: source/uwp/Renderer/idl/AdaptiveCards.Rendering.Uwp.idl [hunk 23]
```diff
@@ -224,12 +229,38 @@ AdaptiveNamespaceStart
     [contract(InternalContract, 1)]
     [internal]
 #endif
-    typedef [v1_enum] enum HAlignment
+    typedef [v1_enum] enum _HorizontalAlignment
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/idl/AdaptiveCards.Rendering.Uwp.idl]: >_ [](start = 27, length = 1)

Lets not include this underscore here or in VerticalAlignment

File: source/shared/cpp/ObjectModel/Enums.cpp [hunk 24]
```diff
@@ -19,6 +19,7 @@ namespace AdaptiveSharedNamespace
             {AdaptiveCardSchemaKey::Attention, "attention"},
             {AdaptiveCardSchemaKey::BackgroundColor, "backgroundColor"},
             {AdaptiveCardSchemaKey::BackgroundImage, "backgroundImage"},
+            {AdaptiveCardSchemaKey::BackgroundImageMode, "mode"},
```
Comment (RebeccaAnne) [inline source/shared/cpp/ObjectModel/Enums.cpp]: >BackgroundImageMode [](start = 36, length = 19)

Since this is just for use to parse things back and forth from json, it would be okay to call this "Mode" and allow it to be reused if something else with a Mode comes along.

File: source/shared/cpp/ObjectModel/ParseUtil.cpp [hunk 25]
```diff
@@ -128,6 +128,33 @@ namespace AdaptiveSharedNamespace
         return propertyValue.asString();
     }
 
+    std::shared_ptr<BackgroundImage> ParseUtil::ExtractBackgroundImage(const Json::Value& json)
```
Comment (RebeccaAnne) [inline source/shared/cpp/ObjectModel/ParseUtil.cpp]: >ExtractBackgroundImage [](start = 48, length = 22)

Maybe just GetBackgroundImage for consistency with the rest of the functions in this file

File: source/uwp/Renderer/lib/AdaptiveCard.cpp [hunk 26]
```diff
@@ -273,7 +275,11 @@ namespace AdaptiveNamespace
         adaptiveCard->SetSpeak(HStringToUTF8(m_speak.Get()));
         adaptiveCard->SetHeight(static_cast<AdaptiveSharedNamespace::HeightType>(m_height));
         adaptiveCard->SetLanguage(HStringToUTF8(m_language.Get()));
-        adaptiveCard->SetBackgroundImage(HStringToUTF8(m_backgroundImage.Get()));
+
+        std::shared_ptr<AdaptiveSharedNamespace::BackgroundImage> sharedBackgroundImage;
+        auto backgroundImage = static_cast<AdaptiveNamespace::AdaptiveBackgroundImage*>(m_backgroundImage.Get());
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/AdaptiveCard.cpp]: >static_cast<AdaptiveNamespace::AdaptiveBackgroundImage*> [](start = 31, length = 56)

Change cast to peek

File: source/uwp/Renderer/lib/AdaptiveContainer.cpp [hunk 27]
```diff
@@ -102,6 +115,11 @@ namespace AdaptiveNamespace
         container->SetStyle(static_cast<AdaptiveSharedNamespace::ContainerStyle>(m_style));
         container->SetVerticalContentAlignment(static_cast<AdaptiveSharedNamespace::VerticalContentAlignment>(m_verticalAlignment));
 
+        std::shared_ptr<AdaptiveSharedNamespace::BackgroundImage> sharedBackgroundImage;
+        auto backgroundImage = static_cast<AdaptiveNamespace::AdaptiveBackgroundImage*>(m_backgroundImage.Get());
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/AdaptiveContainer.cpp]: >static_cast [](start = 31, length = 11)

Change cast to peek

File: source/uwp/Renderer/lib/XamlBuilder.h [hunk 28]
```diff
@@ -39,79 +39,79 @@ namespace AdaptiveNamespace
...
-                                        _In_ ABI::AdaptiveNamespace::IAdaptiveRenderArgs* renderArgs,
-                                        _Outptr_ ABI::Windows::UI::Xaml::IUIElement** choiceSetInputControl);
+            _In_ ABI::AdaptiveNamespace::IAdaptiveRenderContext* renderContext,
+            _In_ ABI::AdaptiveNamespace::IAdaptiveRenderArgs* renderArgs,
+            _Outptr_ ABI::Windows::UI::Xaml::IUIElement** choiceSetInputControl);
         static void BuildDateInput(_In_ ABI::AdaptiveNamespace::IAdaptiveCardElement* adaptiveCardElement,
-                                   _In_ ABI::AdaptiveNamespace::IAdaptiveRenderContext* renderContext,
-                                   _In_ ABI::AdaptiveNamespace::IAdaptiveRenderArgs* renderArgs,
-                                   _Outptr_ ABI::Windows::UI::Xaml::IUIElement** dateInputControl);
+            _In_ ABI::AdaptiveNamespace::IAdaptiveRenderContext* renderContext,
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/XamlBuilder.h]: >             [](start = 0, length = 12)

These should be indented in. Re-run formatting on this file (Ctrl-K Ctrl-D)

File: source/uwp/Renderer/idl/AdaptiveCards.Rendering.Uwp.idl (hunk 23 above)
Comment (carlos-zamora) [reply source/uwp/Renderer/idl/AdaptiveCards.Rendering.Uwp.idl]: So, I don't want to but it get's mad when the underscore isn't there. I get `Error MIDL2003 : redefinition : HorizontalAlignment_Left` and the same for all of the other values. Same with VerticalAlignment. This is because of the Windows::UI::Xaml::HorizontalAlignment.

---
In reply to: [253176897](https://github.com/Microsoft/AdaptiveCards/pull/2316#discussion_r253176897) [](ancestors = 253176897)

File: source/uwp/Renderer/lib/AdaptiveBackgroundImage.cpp (hunk 22 above)
Comment (carlos-zamora) [reply source/uwp/Renderer/lib/AdaptiveBackgroundImage.cpp]: It doesn't work if I make it return an E_INVALIDARG. I think what makes this different is that normally objects are treated as AdaptiveCardElements and so it's weird if you say that you're going to use an Image but you set it to null. Here, every Card, Column, and Container has to have a BackgroundImage by default. In the event that the user doesn't explicitly say what the BackgroundImage is for the element (this case), it should just construct a "basic" background image object. Nothing will be rendered but all the properties are there in case you want to change them. Or should I change it to make it just outright return S_OK and the Card/Column/Container's backgroundImage property is just null?

---
In reply to: [253172646](https://github.com/Microsoft/AdaptiveCards/pull/2316#discussion_r253172646) [](ancestors = 253172646)

File: source/uwp/Renderer/idl/AdaptiveCards.Rendering.Uwp.idl (hunk 23 above)
Comment (RebeccaAnne) [reply source/uwp/Renderer/idl/AdaptiveCards.Rendering.Uwp.idl]: Ugh, that's probably why it was called HAlignment in the first place. Was there a particular reason we changed the name? (other than HAligment being obviously not-ideal) Since this is what's going to be used by API callers i think we really don't want that underscore in there. 

---
In reply to: [254501325](https://github.com/Microsoft/AdaptiveCards/pull/2316#discussion_r254501325) [](ancestors = 254501325,253176897)

File: source/uwp/Renderer/lib/AdaptiveBackgroundImage.cpp (hunk 22 above)
Comment (RebeccaAnne) [reply source/uwp/Renderer/lib/AdaptiveBackgroundImage.cpp]: I'd say if the background image is not set the property should be null. One way to handle this is follow the model of showcardactions which just doesn't try to make the subcard if it's null in the shared model. (AdaptiveShowCardAction::RuntimeClassInitialize)

---
In reply to: [254508406](https://github.com/Microsoft/AdaptiveCards/pull/2316#discussion_r254508406) [](ancestors = 254508406,253172646)

File: source/uwp/Renderer/idl/AdaptiveCards.Rendering.Uwp.idl (hunk 23 above)
Comment (carlos-zamora) [reply source/uwp/Renderer/idl/AdaptiveCards.Rendering.Uwp.idl]: Haha. Really just `HAlignment` being not-ideal. I also figured that we'd want HorizontalAlignment and VerticalAlignment to follow the same naming convention (so not `HAlignment` and `VerticalAlignment`). I can do `HAlignment` and `VAlignment` but I feel like having them be one letter off is kinda a problem. 

---
In reply to: [257419249](https://github.com/Microsoft/AdaptiveCards/pull/2316#discussion_r257419249) [](ancestors = 257419249,254501325,253176897)

File: samples/v1.2/Elements/AdaptiveCard.BackgroundImage.RepeatHorizontally.Center.json [hunk 29]
```diff
@@ -0,0 +1,58 @@
...
+  "type": "AdaptiveCard",
+  "version": "1.2",
+	"backgroundImage": {
+		"url": "https://picsum.photos/60/60?image=101",
+		"mode": "repeatHorizontally",
+		"verticalAlignment": "center"
+	},
+  "body": [
+    {
+      "type": "TextBlock",
```
Comment (RebeccaAnne) [inline samples/v1.2/Elements/AdaptiveCard.BackgroundImage.RepeatHorizontally.Center.json]: >      " [](start = 0, length = 7)

tabs

File: source/uwp/Renderer/lib/AdaptiveColumn.cpp [hunk 30]
```diff
@@ -109,6 +127,17 @@ namespace AdaptiveNamespace
...
+        ComPtr<AdaptiveBackgroundImage> adaptiveBackgroundImage = PeekInnards<AdaptiveBackgroundImage>(m_backgroundImage);
+        std::shared_ptr<AdaptiveSharedNamespace::BackgroundImage> sharedBackgroundImage;
+        if (adaptiveBackgroundImage && SUCCEEDED(adaptiveBackgroundImage->GetSharedModel(sharedBackgroundImage)))
+        {
+            column->SetBackgroundImage(sharedBackgroundImage);
+        }
+        else
+        {
+            column->SetBackgroundImage(nullptr);
+        }
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/AdaptiveColumn.cpp]: Is the else necessary? If you don't set it won't it be null already.
